zrml/global-disputes/src/tests.rs

```diff
@@ -0,0 +1,1360 @@
+        let market_id = 0u128;
+        GlobalDisputes::push_voting_outcome(
+            &market_id,
+            OutcomeReport::Scalar(0),
+            &ALICE,
+            10 * BASE,
+        )
+        .unwrap();
+        GlobalDisputes::push_voting_outcome(
+            &market_id,
+            OutcomeReport::Scalar(20),
+            &ALICE,
+            20 * BASE,
+        )
+        .unwrap();
+        GlobalDisputes::push_voting_outcome(
+            &market_id,
+            OutcomeReport::Scalar(40),
+            &ALICE,
+            30 * BASE,
+        )
+        .unwrap();
+        GlobalDisputes::push_voting_outcome(
+            &market_id,
+            OutcomeReport::Scalar(60),
+            &ALICE,
+            40 * BASE,
+        )
+        .unwrap();
+
+        assert_noop!(
+            GlobalDisputes::vote_on_outcome(
+                Origin::signed(ALICE),
+                market_id,
+                OutcomeReport::Scalar(40),
+                MinOutcomeVoteAmount::get() - 1,
+            ),
+            Error::<Runtime>::AmountTooLow
+        );
+    });
+}
+
+#[test]
+fn determine_voting_winner_sets_the_last_outcome_for_same_vote_balances_as_the_canonical_outcome() {
+    ExtBuilder::default().build().execute_with(|| {
+        let market_id = 0u128;
+        GlobalDisputes::push_voting_outcome(
+            &market_id,
+            OutcomeReport::Scalar(0),
+            &ALICE,
+            10 * BASE,
+        )
+        .unwrap();
+        GlobalDisputes::push_voting_outcome(
+            &market_id,
+            OutcomeReport::Scalar(20),
+            &ALICE,
+            10 * BASE,
+        )
+        .unwrap();
+        GlobalDisputes::push_voting_outcome(
+            &market_id,
+            OutcomeReport::Scalar(40),
+            &ALICE,
+            10 * BASE,
+        )
+        .unwrap();
+        GlobalDisputes::push_voting_outcome(
+            &market_id,
+            OutcomeReport::Scalar(60),
+            &ALICE,
+            10 * BASE,
+        )
+        .unwrap();
+
+        assert_ok!(GlobalDisputes::vote_on_outcome(
+            Origin::signed(ALICE),
+            market_id,
+            OutcomeReport::Scalar(0),
+            42 * BASE
+        ));
+        System::assert_last_event(
+            Event::<Runtime>::VotedOnOutcome {
+                voter: ALICE,
+                market_id,
+                outcome: OutcomeReport::Scalar(0),
+                vote_amount: 42 * BASE,
+            }
+            .into(),
+        );
+        assert_ok!(GlobalDisputes::vote_on_outcome(
+            Origin::signed(BOB),
+            market_id,
+            OutcomeReport::Scalar(20),
+            42 * BASE
+        ));
+        System::assert_last_event(
+            Event::<Runtime>::VotedOnOutcome {
+                voter: BOB,
+                market_id,
+                outcome: OutcomeReport::Scalar(20),
+                vote_amount: 42 * BASE,
+            }
+            .into(),
+        );
```

Comment:
OK

Reply:
I also need to work more with setup functions. The most test were often copy & pastes rather than well thought setup functions. And this made the tests longer and longer, so that each new test requires more think power to know what to copy to build a new test. So in this case it was easier to just add checks to the existing tests. 

So, when I would finally use setup functions, I will write smaller tests and it's much easier to avoid mixed tests.

Reply:
ebd7317f81633b62666b0d1a8f35dc10a771cf92